pkg/injector/injector.go

```diff
@@ -234,7 +249,12 @@ func (i *injector) handleRequest(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		log.Errorf("Can't decode body: %v", err)
 	} else {
-		if !(utils.Contains(i.authUIDs, ar.Request.UserInfo.UID) || utils.Contains(ar.Request.UserInfo.Groups, systemGroup)) {
+		// get a snapshot of the authuids
+		i.authUIDsMu.RLock()
+		authUIDs := i.authUIDs
+		i.authUIDsMu.RUnlock()
+
+		if !(utils.Contains(authUIDs, ar.Request.UserInfo.UID) || utils.Contains(ar.Request.UserInfo.Groups, systemGroup)) {
```

Comment:
My biggest question on the entire PR is here... Does this need a watcher that is constantly notified by the K8s API server if a new SA is available?

Couldn't we just lazily check if the SA exists when we try to use it here?

Yes, this will add a small latency to each operation, but on the other hand doesn't require creating a watcher which takes up a lot more memory.

Reply:
true, depending on the dynamic behavior of the creation/destruction of the virtual clusters, so it is kind of a balancing act, if it is very dynamic it makes sense but if not then not so much. Most of the code here could be repurposed for the evaluation, we would need a garbage collector to prune stale cached sa-ns entries (like a watchdog) if we want to keep a cache to avoid the sa query delay on each pod created for the same vcluster, and only hit the delay the first time (the others it would be cached). let me look into it and check with my peer.

I'll resolve the other comments and get back to this one after. thanks

Reply:
I think a cache may be dangerous because we may be deploying pods with stale information.

IMHO the question here is whether we prioritize latency (removing one API call for each pod being deployed) vs memory usage. I'd personally lean on the second: we are seeing now with the Dapr Operator, for example, how watching for resources can cause significant memory issues on large clusters.

CC: @yaron2

Reply:
I'm leaning on adding another option to default to extra latency, or opting for no latency but more memory usage, SAs are not as pervasive as pods but I understand the issue. The predicate methods can be reused by both after a refactor.

Reply:
just realized that we might have an easy way out. service accounts userinfo should be in the format `system:serviceaccount:(NAMESPACE):(SERVICEACCOUNT)` per https://kubernetes.io/docs/reference/access-authn-authz/authentication/#service-account-tokens and that is the common one. So if we trust the api server that we do, we can just accept this value as a fact and don't query the api server and just try to match against this value, no need for cache nor query.

Reply:
That sounds reasonable to me.

Reply:
For the use case that this was originaly intended (vclusters reusing a single dapr installation) I don't think that it needs to be that dynamic. Creating a watcher for a defined pattern like `vcluster-*` would make total sense to me, instead of adding a check per request. But I might be missing some context about how Dapr work at that level.